pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -146,22 +145,25 @@ func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.
 }
 
 // filterFrom returns a filtered list of instance types based on the provided resource constraints
-func (p *InstanceTypeProvider) filterFrom(instanceTypes []*packing.Instance, constraints *cloudprovider.Constraints, zones []string) []*packing.Instance {
+func (p *InstanceTypeProvider) filterFrom(instanceTypes []*packing.Instance, constraints *AWSConstraints, zones []string) []*packing.Instance {
 	filtered := []*packing.Instance{}
-	architecture := utils.NormalizeArchitecture(constraints.Architecture)
-
 	for _, instanceTypeInfo := range instanceTypes {
-		if (len(constraints.InstanceTypes) == 0 || functional.ContainsString(constraints.InstanceTypes, *instanceTypeInfo.InstanceType)) &&
-			(len(constraints.InstanceTypes) != 0 || p.isDefaultInstanceType(instanceTypeInfo)) &&
-			(architecture == nil || functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.ProcessorInfo.SupportedArchitectures), *architecture)) &&
-			functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.SupportedUsageClasses), "on-demand") &&
-			(len(zones) == 0 || len(functional.IntersectStringSlice(instanceTypeInfo.Zones, zones)) > 0) {
+		if constraints.Constraints == nil ||
+			(p.isInstanceTypeSupported(constraints.InstanceTypes, instanceTypeInfo) &&
+				p.isCapacityTypeSupported(constraints.GetCapacityType(), instanceTypeInfo) &&
+				p.isArchitectureSupported(utils.NormalizeArchitecture(constraints.Architecture), instanceTypeInfo) &&
+				p.isZonesSupported(zones, instanceTypeInfo)) {
 			filtered = append(filtered, instanceTypeInfo)
 		}
 	}
 	return filtered
 }
 
+func (p *InstanceTypeProvider) isInstanceTypeSupported(instanceTypeConstraints []string, instance *packing.Instance) bool {
+	return (len(instanceTypeConstraints) != 0 || p.isDefaultInstanceType(instance)) &&
+		len(instanceTypeConstraints) == 0 || functional.ContainsString(instanceTypeConstraints, *instance.InstanceType)
```

Comment:
This might just be go fmt but the spacing looks off (misleading) here. Are you sure you don't need another set of parens around the 2nd line? (Normally I'm not a big fan of extra parens, but if I don't have to remember the precedence of `||` and `&&` it might be worth it?)

Reply:
As a rule of thumb, if the if statement can't be expressed via a single operator 
```foo && bar && baz```
or 
```foo || bar || baz```
then you should separate the logic into a series of if statements
```
if shortcircuitPredicate {
  return false
}
if otherShortcircuit {
  return true
}
return false

Reply:
fixed